XFormBuilder for easy XForms

kaapstorm opened the pull request:
XFormBuilder is intended to be a simple way to create an XForm, and add questions to it.

It was created for importing a CDISC ODM document as a CommCare app. Third-party solutions lacked important features like nested groups, and most required their input to come via an Excel spreadsheet or JSON, which did not seem to be a good fit. I also took a look at [app_manager.xform.XForm](https://github.com/dimagi/commcare-hq/blob/xform_builder/corehq/apps/app_manager/xform.py#L539-539) but it seemed aimed at getting data from xfoms already built by Vellum. So I thought XFormBuilder might be a useful tool to add. Usage is as simple as

``` python
xform = XFormBuilder('Built by XFormBuilder')
xform.add_question('personal', 'Personal Questions', data_type='group')
xform.add_question('name', 'What is your name?', group='personal')
xform.add_question('fav_color', 'What is your favorite color?',
                   group='personal',
                   choices={'r': 'Red', 'g': 'Green', 'b': 'Blue'})
xml = xform.tostring(pretty_print=True, encoding='utf-8')
```

(Please excuse the ugly XML formatting in the doctest. I can't see why `pretty_print=True` is so not pretty.)

@millerdev @snopoke cc @orangejenny @czue

kaapstorm committed:
XFormBuilder for easy XForms

kaapstorm commented:
(P.S. I think this deserves more than just a doctest. Groups and nested groups don't have tests yet, nor do all the question data types. Unit tests to follow.)

millerdev commented:
What do you think of changing the API to avoid the problem of passing an unknown group name?
```py
xform = XFormBuilder('Built by XFormBuilder')
personal = xform.add_question('personal', 'Personal Questions', data_type='group')
personal.add_question('name', 'What is your name?')
```

kaapstorm commented:
Hmmm. Nice idea! Maybe `add_group` returns a Group object? Differentiating Groups and RepeatGroups would also simplify a couple of lines when adding a question to the form body.

czue commented:
will let others review, but cool stuff!

kaapstorm committed:
Added QuestionGroup class

dimagimon commented on corehq/apps/app_manager/xform_builder.py:
```diff
@@ -0,0 +1,390 @@
+
+            """
+            node_ = etree.Element('input', {'ref': self.get_data_ref(name_, groups_)})
+            node_.append(etree.Element('label', {'ref': "jr:itext('{}')".format(self.get_text_id(name_, groups_))}))
```
E501 line too long (116 > 115 characters)

kaapstorm commented:
Added class for groups, and made XML a little prettier. More tests to follow.

kaapstorm committed:
Support all ODK XForm data types

millerdev commented on corehq/apps/app_manager/xform_builder.py:
```diff
@@ -0,0 +1,398 @@
+            bind = E.bind({'nodeset': self.get_data_ref(name, group),
+                           'type': 'xsd:' + data_type})
+        else:
+            bind = E.bind({'nodeset': self.get_data_ref(name, group)})
```
You could omit this `else` caluse. Binds with only `nodeset` are ignored.

kaapstorm committed:
XSD_TYPES const, ignore binds with only nodeset

kaapstorm committed:
Add question params and allow XML in labels

`repeat_count` question param can be passed a Question instance.
`add_question` menthod renamed to `new_question`, and returns a
Question or QuestionGroup instance.

kaapstorm committed:
XFormBuilder tests

dimagimon commented on corehq/apps/app_manager/xform_builder.py:
```diff
@@ -0,0 +1,432 @@
+"""
+import uuid
+from lxml import etree
+from lxml.builder import E, ElementMaker
```
F401 'ElementMaker' imported but unused

kaapstorm commented:
Added tests. Pending feedback, I'm happy with this now.

millerdev commented:
Test failure looks real: https://s3.amazonaws.com/archive.travis-ci.org/jobs/90074401/log.txt

millerdev commented on corehq/apps/app_manager/xform_builder.py:
```diff
@@ -0,0 +1,432 @@
+... '''    </group>\\n'''
+... '''  </h:body>\\n'''
+... '''</h:html>\\n''')
+True
```
What do you think about converting this doctest into a real unit test? It's hard to tell what went wrong when it fails:
```
Expected:
    True
Got:
    False
```
Also, in general there are [some very good reasons not to use doctests for formal testing](http://nedbatchelder.com/blog/200811/things_i_dont_like_about_doctest.html). I think doctests are well suited to verify that example code (documentation) works, but beyond that they become problematic.

kaapstorm replied:
Good call, thanks @millerdev

millerdev commented on corehq/apps/app_manager/tests/test_xform_builder.py:
```diff
@@ -0,0 +1,104 @@
+        XFormBuilder.new_question should be able to add a group
+        """
+        self.xform.new_question('personal', 'Personal Questions', data_type='group')
+        self.xform.new_question('name', 'What is your name?', group='personal')
```
Is there a reason why this doesn't use `personal.new_group`?

kaapstorm replied:
I test both ways. See two tests down.

millerdev commented on corehq/apps/app_manager/xform_builder.py:
```diff
@@ -0,0 +1,432 @@
+GROUP_TYPES = ('group', 'repeatGroup')  # TODO: Support 'questionList'
+# The subset of ODK data types that are XSD data types
+# cf. http://www.w3.org/TR/xmlschema-2/#built-in-datatypes
+XSD_TYPES = ('string', 'int', 'boolean', 'decimal', 'date', 'time', 'dateTime')
```
Nitpick: you could eliminate a tiny bit of duplication by defining this before `ODK_TYPES` and then
```py
ODK_TYPES = XSD_TYPES + ('select', 'select1', ...
```

kaapstorm replied:
:+1:

kaapstorm committed:
Convert doctest to unit test

kaapstorm committed:
Dedup data types

kaapstorm committed:
OCD

millerdev commented:
I'm satisfied. GTM when tests pass (are we still killing intermediate test runs to reduce the travis backlog?)
